### PR TITLE
fixup json,http protocols

### LINKED ISSUE #326: Issue with new streaming responses
I have a function that failed halfway through, my browser got half the response then it crapped out. Then an error json halfway through:

```html
<!DOCTYPE html>
--
  | <html>
  | <head>
  | <meta charset="UTF-8">
  | <title>My page</title>
  | </head>
  | <body>
  | <h1>My page</h1>
  | <p>This is my app. It may not be the best app, but it&#39;s my app.&lt;br/&gt;&lt;br/&gt;And it&#39;s multilingual!</p>
  | <div>
  | <div><a href="{"error":{"message":"container exit code 1"}}
```

It should have 50X'd back to the user rather than return anything at all. 

```
ERRO[2712] api error                                     action="server.handleRunnerRequest)-fm" app=helloapp code=502 error="container exit code 1"
```

Also, there's no call_id in that log line to pull up the container logs. 

 This should return:

- [ ] 500 internal server error
- [ ] error JSON only

### PR DESCRIPTION
**Pull request by rdallman**
* http now buffers the entire request body from the container before copying
it to the response writer (and sets content length). this is a level of sad i
don't feel comfortable talking about but it is what it is.
* json protocol was buffering the entire body so there wasn't any reason for
us to try to write this directly to the container stdin manually, we needed to
add a bufio.Writer around it anyway it was making too many write(fd) syscalls
with the way it was. this is just easier overall and has the same performance
as http now in my tests, whereas previously this was 50% slower [than http].
* add buffer pool for http & json to share/use. json doesn't create a new
buffer every stinkin request. we need to plumb down content length so that we
can properly size the buffer for json, have to add header size and everything
together but it's probably faster than malloc(); punting on properly sizing.
* json now sets content type to the length of the body from the returned json
blurb from the container

this does not handle imposing a maximum size of the response returned from a
container, which we need to add, but this has been open for some time
(specifically, on json). we can impose this by wrapping the pipes, but there's
some discussion to be had for json specifically we won't be able to just cut
off the output stream and use that (http we can do this). anyway, filing a
ticket...

closes #326 :(((((((

### CODE REVIEW

File: api/agent/protocol/json.go [hunk 1]
```diff
@@ -36,19 +29,19 @@ type CallResponseHTTP struct {
...
 // jsonIn We're not using this since we're writing JSON directly right now, but trying to keep it current anyways, much easier to read/follow
 type jsonIn struct {
-	jsonio
-	CallID      string           `json:"call_id"`
-	ContentType string           `json:"content_type"`
-	Type        string           `json:"type"`
-	Deadline    string           `json:"deadline"`
-	Body        string           `json:"body"`
-	Protocol    *CallRequestHTTP `json:"protocol"`
+	CallID      string          `json:"call_id"`
```
Comment (skinowski) [inline api/agent/protocol/json.go]: :clap:

File: api/agent/protocol/json.go [hunk 2]
```diff
@@ -36,19 +29,19 @@ type CallResponseHTTP struct {
...
+	Type        string          `json:"type"`
+	Deadline    string          `json:"deadline"`
+	Body        string          `json:"body"`
+	ContentType string          `json:"content_type"`
```
Comment (skinowski) [inline api/agent/protocol/json.go]: :clap:

File: api/agent/protocol/json.go [hunk 3]
```diff
@@ -36,19 +29,19 @@ type CallResponseHTTP struct {
...
+	Protocol    CallRequestHTTP `json:"protocol"`
 }
 
 // jsonOut the expected response from the function container
 type jsonOut struct {
-	jsonio
-	Protocol *CallResponseHTTP `json:"protocol,omitempty"`
+	Body        string            `json:"body"`
+	ContentType string            `json:"content_type"`
```
Comment (skinowski) [inline api/agent/protocol/json.go]: :clap:

File: api/agent/protocol/http.go [hunk 4]
```diff
@@ -2,15 +2,23 @@ package protocol
...
+	"sync"
 
 	"github.com/fnproject/fn/api/models"
 	opentracing "github.com/opentracing/opentracing-go"
 )
 
+var (
+	bufPool = &sync.Pool{New: func() interface{} { return new(bytes.Buffer) }}
+)
+
```
Comment (skinowski) [inline api/agent/protocol/http.go]: maybe factory.go is a better place? not a big deal...

File: api/agent/protocol/json.go [hunk 5]
```diff
@@ -62,131 +55,33 @@ func (p *JSONProtocol) IsStreamable() bool {
...
-			return err
-		}
-		err = stdin.Encode(ci.Headers())
-	}
-	err = writeString(err, h.in, "}")
-
-	// close
-	err = writeString(err, h.in, "\n}\n\n")
-	return err
+	return json.NewEncoder(h.in).Encode(in)
```
Comment (skinowski) [inline api/agent/protocol/json.go]: :clap:

**Comment by rdallman** [comment]
thanks. wrt the buffer pool... we have 2 in func_logger as it is and this one as well. it seems like it would be a useful thing to have a general purpose buffer pool that can dispense based on some expected size. for this one, it's not perfect since http is using it for container output and json for request input, but in a heterogeneous workload we can't expect that those 2 differ a whole lot (one of the func_logger ones is just for lines, which should be short generally).

**Comment by skinowski** [comment]
So, with this, we are always buffering. (no streaming mode), which is fine in my opinion. Only to-do here is to set a system wide limit and if resp exceeds, cut off response, get rid of buffered data (clean pipes), then return an error (eg. data too large), right?

**Comment by rdallman** [comment]
something like that. opened #773 - default needs to be handled, as well. we could either return an error or return partial output and somehow try to indicate that their output was too long. not sure how worth it / challenging the partial output thing is (we're truncating logs, but stdout is a little different).

**Pull request merged**